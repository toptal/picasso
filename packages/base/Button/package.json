{
  "name": "@toptal/picasso-button",
  "version": "2.0.1",
  "description": "Toptal UI components library - Button",
  "publishConfig": {
    "access": "public"
  },
  "main": "./dist-package/src/index.js",
  "module": "./dist-package/src/index.js",
  "scripts": {
    "build:package": "tsc -b tsconfig.json",
    "prepublishOnly": "yarn build:package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/toptal/picasso.git"
  },
  "author": "Toptal",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/toptal/picasso/issues"
  },
  "homepage": "https://github.com/toptal/picasso/tree/master/packages/picasso#readme",
  "dependencies": {
    "@toptal/picasso-checkbox": "2.0.1",
    "@toptal/picasso-container": "1.0.3",
    "@toptal/picasso-dropdown": "2.0.3",
    "@toptal/picasso-icons": "1.1.1",
    "@toptal/picasso-loader": "1.0.3",
    "@toptal/picasso-radio": "2.0.1",
    "@toptal/picasso-shared": "15.0.0",
    "@toptal/picasso-utils": "1.0.3",
    "@toptal/picasso-link": "1.0.3",
    "ap-style-title-case": "^1.1.2",
    "@mui/base": "5.0.0-beta.40",
    "classnames": "^2.5.1",
    "tailwind-merge": "^2.2.2"
  },
  "sideEffects": [
    "**/styles.ts",
    "**/styles.js"
  ],
  "peerDependencies": {
    "@toptal/picasso-tailwind": ">=2.5.0",
    "@toptal/picasso-provider": "*",
    "react": ">=16.12.0 < 19.0.0"
  },
  "exports": {
    ".": "./dist-package/src/index.js"
  },
  "devDependencies": {
    "@toptal/picasso-provider": "4.2.1",
    "@toptal/picasso-test-utils": "1.1.1"
  },
  "files": [
    "dist-package/**",
    "!dist-package/tsconfig.tsbuildinfo",
    "src"
  ]
}
