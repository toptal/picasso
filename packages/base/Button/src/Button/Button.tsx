import type { ReactNode, ReactElement, MouseEvent, ElementType } from 'react'
import React, { forwardRef } from 'react'
import cx from 'classnames'
import type {
  StandardProps,
  SizeType,
  ButtonOrAnchorProps,
  OverridableComponent,
  TextLabelProps,
  Classes,
} from '@toptal/picasso-shared'
import { useTitleCase } from '@toptal/picasso-shared'
import type { Theme } from '@material-ui/core'
import { makeStyles, ButtonBase } from '@material-ui/core'
import { Loader } from '@toptal/picasso-loader'
import { Container } from '@toptal/picasso-container'
import { noop, toTitleCase } from '@toptal/picasso-utils'
// we need to ensure the correct order of styles import
// TODO: [FX-4614] To be removed when Link component is migrated to tailwind
import { Link } from '@toptal/picasso-link'

import styles from './styles'

// HACK: This statement is only used to prevent webpack from tree shaking the import
void Link

const useStyles = makeStyles<Theme, Props>(styles, {
  name: 'PicassoButton',
})

export type VariantType =
  | 'primary'
  | 'negative'
  | 'positive'
  | 'secondary'
  | 'transparent'

export type IconPositionType = 'left' | 'right'

export interface Props
  extends StandardProps,
    TextLabelProps,
    ButtonOrAnchorProps {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  as?: ElementType
  /** Disables button */
  disabled?: boolean
  /** Content of Button component */
  children?: ReactNode
  focused?: boolean
  /** Take the full width of a container */
  fullWidth?: boolean
  /** Set hovered style for the button */
  hovered?: boolean
  /** Add an `<Icon />` along Button's children */
  icon?: ReactElement
  /** Icon can be positioned on the left or right */
  iconPosition?: IconPositionType
  /** Shows a loading indicator and disables click events */
  loading?: boolean
  /** Callback invoked when component is clicked */
  onClick?: (event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>) => void
  /** A button can have different sizes */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** The variant to use */
  variant?: VariantType
  /** HTML Value of Button component */
  value?: string | number
  /** HTML title of Button component */
  title?: string
  /** HTML type of Button component */
  type?: 'button' | 'reset' | 'submit'
}

const getClickHandler = (loading?: boolean, handler?: Props['onClick']) =>
  loading ? noop : handler

const getIcon = (
  classes: Classes,
  children: ReactNode,
  icon?: ReactElement,
  iconPosition?: IconPositionType
) => {
  if (!icon) {
    return null
  }

  const {
    icon: iconClass,
    iconLeft: iconLeftClass,
    iconRight: iconRightClass,
  } = classes

  return React.cloneElement(icon, {
    className: cx(iconClass, icon.props.className, {
      [iconLeftClass]: children && iconPosition === 'left',
      [iconRightClass]: children && iconPosition === 'right',
    }),
    key: 'button-icon',
  })
}

export const Button: OverridableComponent<Props> = forwardRef<
  HTMLButtonElement,
  Props
>(function Button(props, ref) {
  const {
    icon,
    iconPosition,
    loading,
    children,
    className,
    style,
    fullWidth,
    variant = 'primary',
    size = 'medium',
    focused,
    hovered,
    disabled,
    active,
    onClick,
    title,
    value,
    type,
    as = 'button',
    titleCase: propsTitleCase,
    ...rest
  } = props
  const classes = useStyles(props)

  const {
    root: rootClass,
    hidden: hiddenClass,
    loader: loaderClass,
    content: contentClass,
  } = classes

  const titleCase = useTitleCase(propsTitleCase)

  const finalChildren = [titleCase ? toTitleCase(children) : children]

  if (icon) {
    const iconComponent = getIcon(classes, children, icon, iconPosition)

    if (iconPosition === 'left') {
      finalChildren.unshift(iconComponent)
    } else {
      finalChildren.push(iconComponent)
    }
  }

  const variantClassName = classes[variant]
  const sizeClassName = classes[size]

  const rootClassName = cx(
    {
      [classes.fullWidth]: fullWidth,
      [classes.active]: active,
      [classes.focused]: focused,
      [classes.hovered]: hovered,
      [classes.disabled]: disabled,
    },
    sizeClassName,
    variantClassName,
    rootClass
  )

  return (
    <ButtonBase
      {...rest}
      ref={ref}
      classes={{
        root: rootClassName,
        focusVisible: cx(classes.focusVisible),
      }}
      onClick={getClickHandler(loading, onClick)}
      className={className}
      style={style}
      disabled={disabled}
      title={title}
      value={value}
      type={type}
      component={as}
      data-component-type='button'
    >
      <Container
        as='span'
        inline
        flex
        direction='row'
        alignItems='center'
        className={cx({ [hiddenClass]: loading }, contentClass)}
      >
        {finalChildren}
      </Container>

      {loading && (
        <Loader variant='inherit' className={loaderClass} inline size='small' />
      )}
    </ButtonBase>
  )
})

Button.defaultProps = {
  active: false,
  as: 'button',
  children: null,
  disabled: false,
  focused: false,
  fullWidth: false,
  hovered: false,
  iconPosition: 'left',
  loading: false,
  onClick: noop,
  size: 'medium',
  type: 'button',
  variant: 'primary',
}

Button.displayName = 'Button'

export default Button
