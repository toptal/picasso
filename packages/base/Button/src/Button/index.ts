import type { Props } from './Button'
export { default as Button } from './Button'
export type ButtonProps = Props
export * from './Button'
export * from './styles'
