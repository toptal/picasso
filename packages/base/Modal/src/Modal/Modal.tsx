import type { ReactNode, HTMLAttributes } from 'react'
import { Modal as Dialog } from '@mui/base/Modal'
import React, {
  forwardRef,
  useEffect,
  useRef,
  useCallback,
  useContext,
} from 'react'
import cx from 'classnames'
import type {
  BaseProps,
  SizeType,
  TransitionProps,
} from '@toptal/picasso-shared'
import { usePicassoRoot, RootContext } from '@toptal/picasso-provider'
import { CloseMinor16 } from '@toptal/picasso-icons'
import {
  useCombinedRefs,
  ModalManager,
  usePageScrollLock,
} from '@toptal/picasso-utils'
import { ButtonCircular } from '@toptal/picasso-button'
import { Fade } from '@toptal/picasso-fade'
import { Backdrop } from '@toptal/picasso-backdrop'
import ModalContext from '@toptal/picasso-modal-context'

import { ModalPaper } from '../ModalPaper'

type ContainerValue = HTMLElement | (() => HTMLElement)
type Alignment = 'top' | 'centered'

export interface Props extends BaseProps, HTMLAttributes<HTMLDivElement> {
  /** Content of Modal component */
  children: ReactNode
  /** Whether modal should be displayed */
  open: boolean
  /** Width of modal */
  size?:
    | SizeType<'xsmall' | 'small' | 'medium' | 'large' | 'xlarge'>
    | 'full-screen'
  /** Callback executed when backdrop was clicked */
  onBackdropClick?: () => void
  /** If `true`, clicking the backdrop will not fire `onClose` or `onBackdropClick` */
  disableBackdropClick?: boolean
  /** Callback executed when attempting to close modal */
  onClose?: () => void
  /** Callback executed when modal is being opened */
  onOpen?: () => void
  /** A node, or a function that returns node. The container will have the portal children appended to it. */
  container?: ContainerValue
  /** If `true`, the backdrop is not rendered */
  hideBackdrop?: boolean
  /** Position of the modal relative to the browser's viewport */
  align?: Alignment
  /** Animation lifecycle callbacks. Backed by [react-transition-group/Transition](https://reactcommunity.org/react-transition-group/transition#Transition-props) */
  transitionProps?: TransitionProps
  transitionDuration?: number
  /** used for specifying aria attributes, changing role, or customizing styles */
  paperProps?: React.HTMLAttributes<HTMLDivElement>
  testIds?: {
    closeButton?: string
  }
}

const defaultManager = new ModalManager()

// https://github.com/udacity/ud891/blob/gh-pages/lesson2-focus/07-modals-and-keyboard-traps/solution/modal.js#L25
// found in https://developers.google.com/web/fundamentals/accessibility/focus/using-tabindex
const focusableElementsString =
  'a[href], area[href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), button:not([disabled]), iframe, object, embed, [tabindex="0"], [contenteditable]'
const tooltipContainerString = '[role=tooltip]'

const focusFirstFocusableElement = (node: Element) => {
  const elements = node.querySelectorAll(focusableElementsString)
  // Convert NodeList to Array
  const focusableElements = Array.prototype.slice.call(elements)

  if (focusableElements.length > 0) {
    focusableElements[0].focus()
  }
}

const isFocusInsideModal = (modalNode: Element) => {
  const modalContainsFocusedElement = modalNode.contains(document.activeElement)

  if (modalContainsFocusedElement) {
    return true
  }

  return false
}

const isFocusInsideTooltip = () => {
  const tooltipContainers = document.querySelectorAll(tooltipContainerString)

  if (tooltipContainers.length === 0) {
    return false
  }

  const tooltipContainsFocusedElement = Array.from(tooltipContainers).some(
    container => container.contains(document.activeElement)
  )

  if (tooltipContainsFocusedElement) {
    return true
  }

  return false
}

const generateKey = (() => {
  let count = 0

  return () => ++count
})()

// eslint-disable-next-line react/display-name
export const Modal = forwardRef<HTMLDivElement, Props>(function Modal(
  props,
  ref
) {
  const {
    children,
    open,
    size = 'medium',
    onBackdropClick,
    onClose,
    onOpen,
    className,
    style,
    container,
    hideBackdrop = false,
    transitionDuration = 300,
    paperProps,
    align = 'centered',
    testIds,
    transitionProps,
    disableBackdropClick = false,
    ...rest
  } = props
  const picassoRootContainer = usePicassoRoot()
  const modalRef = useCombinedRefs<HTMLDivElement>(
    ref,
    useRef<HTMLDivElement>(null)
  )
  const modalId = useRef(generateKey())
  const { rootRef } = useContext(RootContext)

  useEffect(() => {
    const handleDocumentFocus = () => {
      if (!rootRef?.current) {
        console.warn(
          'Modal is not rendered inside PicassoRoot, some things might not work as expected. Please open the Modal on mount using useEffect.'
        )
      }

      if (!defaultManager.isTopModal(modalId.current)) {
        return
      }

      if (!modalRef || !modalRef.current) {
        return
      }

      if (isFocusInsideModal(modalRef.current)) {
        return
      }

      if (isFocusInsideTooltip()) {
        return
      }

      focusFirstFocusableElement(modalRef.current)
    }

    if (!open) {
      return
    }

    document.addEventListener('focus', handleDocumentFocus, true)

    return () => {
      document.removeEventListener('focus', handleDocumentFocus, true)
    }
  }, [open, modalRef, rootRef])

  useEffect(() => {
    const currentModalId = modalId.current

    if (open) {
      defaultManager.add(currentModalId)
    }

    return () => {
      defaultManager.remove(currentModalId)
    }
  }, [open])

  usePageScrollLock(open)

  const handleClose = useCallback(
    (_event, reason: 'backdropClick' | 'escapeKeyDown') => {
      if (reason === 'escapeKeyDown' && onClose) {
        onClose()
      } else if (reason === 'backdropClick' && !disableBackdropClick) {
        if (onBackdropClick) {
          onBackdropClick()
        }

        if (onClose) {
          onClose()
        }
      }
    },
    [disableBackdropClick, onBackdropClick, onClose]
  )

  const duration = transitionProps?.timeout || transitionDuration
  const backdropProps = { transitionDuration: duration }

  return (
    <Dialog
      {...rest}
      ref={modalRef}
      className={cx(
        className,
        'fixed z-modal inset-0 flex flex-col text-lg leading-[normal] justify-center items-center'
      )}
      style={style}
      slots={{
        backdrop: Backdrop,
      }}
      closeAfterTransition
      slotProps={{
        backdrop: backdropProps,
      }}
      container={container || picassoRootContainer}
      hideBackdrop={hideBackdrop}
      onClose={handleClose}
      open={open}
      disableEnforceFocus // we need our own mechanism to keep focus inside the Modals
      disableScrollLock
    >
      <Fade
        in={open}
        onEnter={onOpen}
        onExited={transitionProps?.onExited}
        timeout={transitionDuration}
      >
        <ModalPaper size={size} align={align} tabIndex={-1} {...paperProps}>
          <ModalContext.Provider value>
            {children}
            {onClose && (
              <ButtonCircular
                aria-label='Close'
                variant='flat'
                className='absolute top-8 right-8'
                onClick={onClose}
                data-testid={testIds?.closeButton}
              >
                <CloseMinor16 />
              </ButtonCircular>
            )}
          </ModalContext.Provider>
        </ModalPaper>
      </Fade>
    </Dialog>
  )
})

Modal.defaultProps = {
  hideBackdrop: false,
  disableBackdropClick: false,
  size: 'medium',
  transitionDuration: 300,
  align: 'centered',
}

Modal.displayName = 'Modal'

export default Modal
