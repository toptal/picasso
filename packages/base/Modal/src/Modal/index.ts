import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Modal'

export { default as Modal } from './Modal'
export { default as ModalContext } from '@toptal/picasso-modal-context'
export type ModalProps = OmitInternalProps<Props>
export * from './Modal'
