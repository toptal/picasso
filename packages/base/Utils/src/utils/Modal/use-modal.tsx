import { useCallback, useState } from 'react'

const useModal = () => {
  const [isOpen, setIsOpen] = useState(false)

  const hideModal = useCallback(() => setIsOpen(false), [setIsOpen])

  const showModal = useCallback(() => setIsOpen(true), [setIsOpen])

  return {
    showModal,
    hideModal,
    isOpen,
  }
}

export { useModal }
