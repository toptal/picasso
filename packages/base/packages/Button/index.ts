/* eslint-disable import/no-extraneous-dependencies */
import type { Props } from './Button'
export { default } from './Button'
export type ButtonProps = Props
export * from './Button'
