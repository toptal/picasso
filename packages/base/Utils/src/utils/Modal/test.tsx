import React from 'react'
import {
  render,
  fireEvent,
  waitForElementToBeRemoved,
} from '@toptal/picasso/test-utils'

import Button from '../../Button'
import Modal from '../../Modal'
import { useModal } from './use-modal'

let modalRoot: HTMLElement

describe('Modal', () => {
  beforeAll(() => {
    modalRoot = document.createElement('div')

    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
  })

  it('useModal opens and closes modal', async () => {
    const TestComponent = () => {
      const { showModal, hideModal, isOpen } = useModal()

      return (
        <>
          <Button onClick={showModal}>Show</Button>

          <Modal open={isOpen}>
            <p>Modal content</p>
            <Button onClick={hideModal}>Hide</Button>
          </Modal>
        </>
      )
    }

    const { getByText, queryByText, baseElement } = render(<TestComponent />)

    const showModalButton = getByText('Show')

    fireEvent.click(showModalButton)

    expect(queryByText('Modal content')).toBeTruthy()
    expect(baseElement).toMatchSnapshot()

    const hideModalButton = getByText('Hide')

    fireEvent.click(hideModalButton)
    await waitForElementToBeRemoved(() => getByText('Hide'))

    expect(queryByText('Modal content')).toBeFalsy()
    expect(baseElement).toMatchSnapshot()
  })

  it('useModal shows multiple modals at the same time', () => {
    const TestComponent = () => {
      const { showModal: showModal1, isOpen: isOpen1 } = useModal()
      const { showModal: showModal2, isOpen: isOpen2 } = useModal()

      return (
        <>
          <Button onClick={showModal1}>Show first</Button>
          <Modal open={isOpen1}>
            <p>First modal content</p>
          </Modal>

          <Button onClick={showModal2}>Show second</Button>
          <Modal open={isOpen2}>
            <p>Second modal content</p>
          </Modal>
        </>
      )
    }

    const { getByText, queryByText, baseElement } = render(<TestComponent />)

    fireEvent.click(getByText('Show first'))
    fireEvent.click(getByText('Show second'))

    expect(queryByText('First modal content')).toBeTruthy()
    expect(queryByText('Second modal content')).toBeTruthy()

    expect(baseElement).toMatchSnapshot()
  })
})
