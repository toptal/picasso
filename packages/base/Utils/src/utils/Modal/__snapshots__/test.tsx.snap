// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal useModal opens and closes modal 1`] = `
<body>
  <div
    id="modal-root"
  />
  <div>
    <div
      class="Picasso-root"
    >
      <button
        aria-disabled="false"
        aria-hidden="true"
        class="base-Button text-lg inline-flex items-center justify-center select-none appearance-none m-0 relative normal-case align-middle transition-colors duration-350 ease-out shrink-0 outline-none [&+&]:ml-4 cursor-pointer no-underline hover:no-underline rounded-sm shadow-none focus-visible:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] focus-within:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] border-none text-white visited:text-white hover:bg-[#4269D6] active:bg-[#1A41AB] bg-blue min-w h-8 py-0 px-4"
        data-component-type="button"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline font-semibold whitespace-nowrap text-button"
        >
          Show
        </span>
      </button>
      <div
        class="MuiDialog-root PicassoModal-root"
        role="presentation"
        style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
      >
        <div
          aria-hidden="true"
          class="MuiBackdrop-root"
          style="opacity: 1; webkit-transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        />
        <div
          data-test="sentinelStart"
          tabindex="0"
        />
        <div
          class="MuiDialog-container PicassoModal-container MuiDialog-scrollPaper"
          role="none presentation"
          style="opacity: 1; webkit-transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          tabindex="-1"
        >
          <div
            class="MuiPaper-root MuiDialog-paper PicassoModal-paper PicassoModal-medium MuiDialog-paperScrollPaper MuiDialog-paperWidthFalse MuiPaper-elevation2"
            role="dialog"
          >
            <p>
              Modal content
            </p>
            <button
              aria-disabled="false"
              class="base-Button text-lg inline-flex items-center justify-center select-none appearance-none m-0 relative normal-case align-middle transition-colors duration-350 ease-out shrink-0 outline-none [&+&]:ml-4 cursor-pointer no-underline hover:no-underline rounded-sm shadow-none focus-visible:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] focus-within:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] border-none text-white visited:text-white hover:bg-[#4269D6] active:bg-[#1A41AB] bg-blue min-w h-8 py-0 px-4"
              data-component-type="button"
              role="button"
              tabindex="0"
              type="button"
            >
              <span
                class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline font-semibold whitespace-nowrap text-button"
              >
                Hide
              </span>
            </button>
          </div>
        </div>
        <div
          data-test="sentinelEnd"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Modal useModal opens and closes modal 2`] = `
<body>
  <div
    id="modal-root"
  />
  <div>
    <div
      class="Picasso-root"
    >
      <button
        aria-disabled="false"
        class="base-Button text-lg inline-flex items-center justify-center select-none appearance-none m-0 relative normal-case align-middle transition-colors duration-350 ease-out shrink-0 outline-none [&+&]:ml-4 cursor-pointer no-underline hover:no-underline rounded-sm shadow-none focus-visible:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] focus-within:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] border-none text-white visited:text-white hover:bg-[#4269D6] active:bg-[#1A41AB] bg-blue min-w h-8 py-0 px-4"
        data-component-type="button"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline font-semibold whitespace-nowrap text-button"
        >
          Show
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`Modal useModal shows multiple modals at the same time 1`] = `
<body>
  <div
    id="modal-root"
  />
  <div>
    <div
      class="Picasso-root"
    >
      <button
        aria-disabled="false"
        aria-hidden="true"
        class="base-Button text-lg inline-flex items-center justify-center select-none appearance-none m-0 relative normal-case align-middle transition-colors duration-350 ease-out shrink-0 outline-none [&+&]:ml-4 cursor-pointer no-underline hover:no-underline rounded-sm shadow-none focus-visible:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] focus-within:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] border-none text-white visited:text-white hover:bg-[#4269D6] active:bg-[#1A41AB] bg-blue min-w h-8 py-0 px-4"
        data-component-type="button"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline font-semibold whitespace-nowrap text-button"
        >
          Show first
        </span>
      </button>
      <button
        aria-disabled="false"
        aria-hidden="true"
        class="base-Button text-lg inline-flex items-center justify-center select-none appearance-none m-0 relative normal-case align-middle transition-colors duration-350 ease-out shrink-0 outline-none [&+&]:ml-4 cursor-pointer no-underline hover:no-underline rounded-sm shadow-none focus-visible:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] focus-within:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] border-none text-white visited:text-white hover:bg-[#4269D6] active:bg-[#1A41AB] bg-blue min-w h-8 py-0 px-4"
        data-component-type="button"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline font-semibold whitespace-nowrap text-button"
        >
          Show second
        </span>
      </button>
      <div
        aria-hidden="true"
        class="MuiDialog-root PicassoModal-root"
        role="presentation"
        style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
      >
        <div
          aria-hidden="true"
          class="MuiBackdrop-root"
          style="opacity: 1; webkit-transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        />
        <div
          data-test="sentinelStart"
          tabindex="0"
        />
        <div
          class="MuiDialog-container PicassoModal-container MuiDialog-scrollPaper"
          role="none presentation"
          style="opacity: 1; webkit-transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          tabindex="-1"
        >
          <div
            class="MuiPaper-root MuiDialog-paper PicassoModal-paper PicassoModal-medium MuiDialog-paperScrollPaper MuiDialog-paperWidthFalse MuiPaper-elevation2"
            role="dialog"
          >
            <p>
              First modal content
            </p>
          </div>
        </div>
        <div
          data-test="sentinelEnd"
          tabindex="0"
        />
      </div>
      <div
        class="MuiDialog-root PicassoModal-root"
        role="presentation"
        style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
      >
        <div
          aria-hidden="true"
          class="MuiBackdrop-root"
          style="opacity: 1; webkit-transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        />
        <div
          data-test="sentinelStart"
          tabindex="0"
        />
        <div
          class="MuiDialog-container PicassoModal-container MuiDialog-scrollPaper"
          role="none presentation"
          style="opacity: 1; webkit-transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          tabindex="-1"
        >
          <div
            class="MuiPaper-root MuiDialog-paper PicassoModal-paper PicassoModal-medium MuiDialog-paperScrollPaper MuiDialog-paperWidthFalse MuiPaper-elevation2"
            role="dialog"
          >
            <p>
              Second modal content
            </p>
          </div>
        </div>
        <div
          data-test="sentinelEnd"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</body>
`;
